src/meta/meta_ckpt.c

```diff
@@ -918,9 +922,18 @@ __wt_meta_sysinfo_set(WT_SESSION_IMPL *session)
         WT_ERR(__wt_metadata_update(session, WT_SYSTEM_CKPT_URI, buf->data));
     }
 
-    /* We also need to record the oldest timestamp in the metadata so we can set it on startup. */
+    /*
+     * We also need to record the oldest timestamp in the metadata so we can set it on startup. We
+     * should set the checkpoint's oldest timestamp as the minimum of the current oldest timestamp
+     * and the checkpoint timestamp.
+     *
+     * Cache the oldest timestamp to prevent us from reading two different values.
+     */
+    oldest_timestamp = txn_global->oldest_timestamp;
+    /* Add a read barrier to prevent the oldest_timestamp from being optimized away. */
```

Comment:
It's not just being optimized away, it's that along with being read twice with different values.